tla-assignments/src/main/scala/at/forsyte/apalache/tla/assignments/SmtFreeSymbolicTransitionExtractor.scala

```diff
@@ -0,0 +1,174 @@
+          if ( s.isEmpty ) None
+          else {
+            val locString = getLocString( childEx )
+            Some( s"$locString - Missing assignments to: ${s.mkString( "," )}" )
+          }
+      }
+
+      throw new AssignmentException( problems.mkString("\n") )
+    }
+    // Assuming all branches assign the same variables, works correctly if args.isEmpty (for whatever reason)
+    val unifiedStrat = (childStates map { _.partialStrat }).foldLeft( s.partialStrat ) { _ ++ _ }
+    PartialState( unassignedEverywhere, unifiedStrat )
+  }
+
+  // Transition method between partial states
+  private def getStrategyOptInternal( s: PartialState, letInOperMap: BodyMap ): TlaEx => PartialState = {
+    /** Base case, assignment candidates  */
+    case ex@OperEx( TlaOper.eq, OperEx( TlaActionOper.prime, NameEx( name ) ), star ) =>
+      // First, check if star contains unassigned varaible access
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      // if `name` is not yet assigned, this spot becomes an assignment
+      if ( s.unassignedVars.contains( name )) PartialState( s.unassignedVars - name, s.partialStrat :+ ex.ID )
+      else s
+
+    /** Base case, manual assignments */
+    case ex@OperEx( BmcOper.assign, OperEx( TlaActionOper.prime, NameEx( name ) ), star ) =>
+      // Similar to above, except manual assignments throw if spurious
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      if ( s.unassignedVars.contains( name )) PartialState( s.unassignedVars - name, s.partialStrat :+ ex.ID )
+      else {
+        val locString = getLocString( ex )
+        throw new AssignmentException( s"Manual assignment at $locString is spurious: $name is already assigned!" )
+      }
+
+    /** Conjunction */
+    case OperEx( TlaBoolOper.and, args@_* ) =>
+      // We sequentially update the partial state
+      args.foldLeft( s ) { getStrategyOptInternal(_, letInOperMap)(_) }
+    /** Disjunction */
+    case ex@OperEx( TlaBoolOper.or, args@_* ) =>
+      handleDisjunctionOrITE( s, letInOperMap,  args)
+
+    /** \E quantifier */
+    case OperEx( TlaBoolOper.exists, NameEx( _ ), star, subEx ) =>
+      // We need to check that star does not contain unassigned primes.
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      // if no exception is thrown, continue in setEx
+      getStrategyOptInternal( s, letInOperMap )( subEx )
+
+
+    /** ITE */
+    case OperEx( TlaControlOper.ifThenElse, star, thenAndElse@_* ) =>
+      // We need to check that star does not contain unassigned primes.
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      // The rest is equivalent to the disjunction case
+      handleDisjunctionOrITE( s, letInOperMap, thenAndElse)
+
+    /** Nullary LetIn */
+    case LetInEx( body, defs@_* ) =>
+      // Sanity check, all operators must be nullary and non-recursive
+      assert( defs.forall { d => d.formalParams.isEmpty && !d.isRecursive } )
```

Comment:
I assume we just assert here because there are earlier checks that should guarantee this property by the time we get to this part of the process?

Reply:
Correct. After inlining, depending on an option, we either end up with no let-in operators at all, or with only nullary ones.